tsdate/prior.py

```diff
@@ -991,16 +1006,129 @@ def fill_priors(node_parameters, timepoints, ts, Ne, *, prior_distr, progress=Fa
     return prior_times
 
 
+class MixturePrior:
+    """
+    Maps ConditionalCoalescentPrior onto nodes in a tree sequence and creates
+    time-discretized priors
+    """
+
+    def __init__(
+        self,
+        tree_sequence,
+        approximate_priors=False,
+        approx_prior_size=None,
+        prior_distribution="lognorm",
+        allow_unary=False,
+        progress=False,
+    ):
+
+        if approximate_priors:
+            if not approx_prior_size:
+                approx_prior_size = 1000
+        else:
+            if approx_prior_size is not None:
+                raise ValueError(
+                    "Can't set approx_prior_size if approximate_prior is False"
+                )
+
+        contmpr_ts, node_map = util.reduce_to_contemporaneous(tree_sequence)
+        if contmpr_ts.num_nodes != tree_sequence.num_nodes:
+            raise ValueError(
+                "Passed tree sequence is not simplified and/or contains "
+                "noncontemporaneous samples"
+            )
+        span_data = SpansBySamples(
+            contmpr_ts, progress=progress, allow_unary=allow_unary
+        )
+
+        base_priors = ConditionalCoalescentTimes(
+            approx_prior_size, prior_distribution, progress=progress
+        )
+
+        base_priors.add(contmpr_ts.num_samples, approximate_priors)
+        for total_fixed in span_data.total_fixed_at_0_counts:
+            # For missing data: trees vary in total fixed node count =>
+            # have different priors
+            if total_fixed > 0:
+                base_priors.add(total_fixed, approximate_priors)
+        prior_params_contmpr = base_priors.get_mixture_prior_params(span_data)
+
+        # Map the nodes in the prior params back to the node ids in the original ts
+        self.prior_params = prior_params_contmpr[node_map, :]
+        self.base_priors = base_priors
+        self.tree_sequence = tree_sequence
+        self.prior_distribution = prior_distribution
+
+    def make_discretized_prior(self, population_size, timepoints=20, progress=False):
+        """
+        Calculate prior grid for a set of timepoints and a population size history
+        """
+
+        if isinstance(population_size, np.ndarray):
+            if population_size.ndim != 2:
+                raise ValueError("Array 'population_size' must be two-dimensional")
+            if population_size.shape[1] != 2:
+                raise ValueError(
+                    "Population size array must have two columns that contain \
+                    epoch start times and population sizes, respectively"
```

Comment:
"Epoch start" is in reverse time (i.e. the most recent time in the epoch), so using the word "start" might confuse anyone working with real data who doesn't think in reverse time.